chips/stm32f303xc/src/flash.rs

```diff
@@ -0,0 +1,608 @@
+                        self.registers.cr.modify(Control::PG::CLEAR);
+                        self.state.set(FlashState::Ready);
+                        self.write_counter.set(0);
+
+                        self.client.map(|client| {
+                            self.buffer.take().map(|buffer| {
+                                client.write_complete(buffer, hil::flash::Error::CommandComplete);
+                            });
+                        });
+                    } else {
+                        self.program_halfword();
+                    }
+                }
+                FlashState::Erase => {
+                    if self.registers.cr.is_set(Control::PER) {
+                        self.registers.cr.modify(Control::PER::CLEAR);
+                    }
+
+                    if self.registers.cr.is_set(Control::MER) {
+                        self.registers.cr.modify(Control::MER::CLEAR);
+                    }
+
+                    self.state.set(FlashState::Ready);
+                    self.client.map(|client| {
+                        client.erase_complete(hil::flash::Error::CommandComplete);
+                    });
+                }
+                FlashState::WriteOption => {
+                    self.registers.cr.modify(Control::OPTPG::CLEAR);
+                    self.state.set(FlashState::Ready);
+                }
+                FlashState::EraseOption => {
+                    self.registers.cr.modify(Control::OPTER::CLEAR);
+                    self.state.set(FlashState::Ready);
+                }
+                _ => {}
+            }
+        }
+
+        if self.state.get() == FlashState::Read {
+            self.state.set(FlashState::Ready);
+            self.client.map(|client| {
+                self.buffer.take().map(|buffer| {
+                    client.read_complete(buffer, hil::flash::Error::CommandComplete);
+                });
+            });
+        }
+
+        if self.registers.sr.is_set(Status::WRPRTERR) {
+            if self.registers.cr.is_set(Control::PG) {
+                self.registers.cr.modify(Control::PG::CLEAR);
+            }
+
+            if self.registers.cr.is_set(Control::OPTPG) {
+                self.registers.cr.modify(Control::OPTPG::CLEAR);
+            }
+
+            // Reset by writing a 1.
+            self.registers.sr.modify(Status::WRPRTERR::SET);
+            panic!("WRPRTERR: programming a write-protected address.");
+        }
+
+        if self.registers.sr.is_set(Status::PGERR) {
+            if self.registers.cr.is_set(Control::PG) {
+                self.registers.cr.modify(Control::PG::CLEAR);
+            }
+
+            if self.registers.cr.is_set(Control::OPTPG) {
+                self.registers.cr.modify(Control::OPTPG::CLEAR);
+            }
+
+            // Reset by writing a 1.
+            self.registers.sr.modify(Status::PGERR::SET);
+            panic!("PGERR: address contains a value different from 0xFFFF before programming.");
```

Comment:
also confused here by the motivation for modifying all these registers only to panic

Reply:
Forgot to delete the panics. I used them for debugging here.

Reply:
Btw, do you think i should replace them with debugs or should I just delete them outright?

Reply:
Do each of these conditions indicate that the (read/write/erase) operation failed? If so, I would say that it is probably best to deliver an error to the client via the appropriate callback.

Reply:
Yes. Both PGERR and WRPTERR indicate failure. I guess that I will use the write_complete callback as I have done further above.

Reply:
sounds good!

Reply:
While we are here, it would be nice to make the kernel flash Error enum generic so that users could define specific error messages. Here I will have to send the same FlashError via callback for two different issues(page write protection vs. page not being erased before programming)

Reply:
Yeah, that enum definitely needs some work, but making it generic probably doesn't work well with capsules being able to make decisions independent of the underlying driver. Instead we probably just need to add additional, more specific error variants to the enum.